### PR TITLE
[no-master] Fix #2175: Add support for ECMAScript 2015 modules.

### LINKED ISSUE #2175: Integrating with JS module systems
# Goal

Seamlessly integrate Scala.js applications with existing JavaScript module systems.
# Background
### JS module systems

Until quite recently the use of _modules_ in JavaScript was not the common. Instead, all libraries polluted the global namespace with their entrypoints. Nowadays there are several competing module systems including AMD, CommonJS and ES6's own module syntax. Traditionally AMD has been a favorite amongs frontend devs due to its dynamic nature, whereas Node.js users went with CommonJS (as it's natively supported by Node).

Latest in the game is the ES6 module syntax, but as it's the only language level module system, it's bound to become the standard in the future.
### Bundling

JS modules tend to be small and numerous, so it makes sense to bundle them together to create one downloadable JS file (just like what Scala.js does) instead of browser downloading potentially dozens of individual files. The bundling process also allows removal of unused exports, reducing the final code size.

Most common tools for bundling these days are [browserify](http://browserify.org/) and [webpack](https://webpack.github.io/), plus a few others like [JSPM](http://jspm.io/) and [RollUp](http://rollupjs.org/).

Due to bundling, the build process of a modern JS web app is quite close to what happens in Scala.js, including dependency/version management, compilation (TS/ES6 to ES5) and linking (bundling).
### Current Scala.js situation

Scala.js is totally oblivious to JS modules and related bundling activities. It operates in its own sandbox, producing a single JS file that assumes dependencies will be available at runtime through the global namespace. The build process supports simplistic bundling (concatenation) of JS libs, to create a single JS dependencies file to be included as part of the application.

Luckily most frontend JS libs today are available in the "ready-to-use" format that you can directly include in the website (or in Scala.js jsdeps), but recently more and more libs have been published only in the module format (for example [Material UI](https://github.com/callemall/material-ui)). On Node.js CommonJS is ubiquitous, so most libs publish only in the module format. To use these libraries in a Scala.js project, developers need to package them separately using the JS toolchain.
# Solutions
## Simple ad-hoc

A quick-and-dirty way of using JS modules in a Scala.js project is to pre-bundle them using the regular JS tools like `webpack`. All you need to do is to define a short JavaScript file serving as an entry point for bundling (and have the dependencies available in the npm style). For example to use React, you would define following:

``` js
window.React = require('react');
```

Running `webpack` produces a `bundle.js` which can be included as part of `jsDependencies` and it will expose React in the global `window` object so it can be used in the Scala.js application.

**Pros**
- Backwards compatible
- No changes required to Scala.js tool chain
- Simple to implement
- Easy to support different module formats (by choosing proper JS tools)

**Cons**
- Requires use of JS tools such as `npm` and `webpack`
- Doesn't really integrate
- No optimization possible (everything is bundled, regardless of real usage)
- Scala.js exports are not available to JS components
## ES6 module output

Next option is to have the Scala.js tool chain output a compatible ES6 module, which can then be included in the normal ES6 processing chain. Of course we could also do this with CommonJS syntax (which is what the vast majority of JS modules are currently using), but at this point it probably makes more sense to go all-in with ES6 modules.
#### Changes required

For this to work, Scala.js would need to produce a JS file with valid ES6 style `import`/`export` statements. 

As there is currently no _import_ mechanism for JS except the dynamic runtime binding using global namespace, new syntax needs to be introduced. For example to import `React` one would write something like:

``` scala
@JSImport("react")
@js.native
object React extends React

@js.native
trait React extends Object { ... }
```

which would generate JS code like

``` javascript
import __JSImport_React from "react";
```

where `__JSImport_React` is an internal generated name for this import, to be used in the code instead of the usual `$g["React"]`.

Similarly anything that is exported with `@JSExport` would be exposed using the ES6 `export` syntax.

The launcher source would also need to change to use `import` instead of accessing the code directly from the global namespace. This launcher would then be the entrypoint for the JS bundler.

**Pros**
- True support of ES6 modules
- Optimization is feasible due to explicit `import` definitions
- Scala.js exports are available to JS modules in the bundling process
- Backwards compatible (if optional)

**Cons**
- Still requires use of JS tools
  - Most current bundling tools support CommonJS, but not ES6 modules
- Bundling must be performed to actually use the JS generated by Scala.js
- JS tools have _very low_ performance when dealing with large JS files produced by Scala.js (for example running webpack on a 4MB `client-fastopt.js` file takes 10-15 seconds, even though there is nothing for it to do. Browserify is even slower.
- CommonJS modules must be converted to ES6 before bundling (or ES6 modules to CJS)
## Full ES6 module bundler

To fully take advantage of ES6 modules we would need to do the bundling part ourselves. In this solution the Scala.js output would be a true ES6 module (as in the previous solution) but instead of using JS tool chain to bundle the final application, we would have our own tool.
#### Changes required

Module dependencies would need their own definition in the build file (in addition to `jsDependencies`) because module archives need to be fully extracted before bundling. For example:

``` scala
jsModules += "org.webjars.npm" % "react" % "0.14.6"
```

WebJar extraction would need custom implementation, to produce a suitable directory structure for bundling. For example the `react` NPM webjar has a directory structure like `META-INF/resources/webjars/react/0.14.6/` whereas what we really need is just `react`. The dependencies of NPM webjars have to be handled as well, to produce a flat structure. One issue is that NPM WebJar generation is not automatic for dependencies (https://github.com/webjars/webjars/issues/1186), so if dependencies are missing, the developer must manually go and add them on the webjar org website.

Actual bundling logic must be implemented. This involves parsing JS files to produce ASTs, then transforming ASTs, replacing any `import`s with appropriate calls and storing modules in their own scopes. Also any CommonJS modules must first be converted into ES6 format.

**Pros**
- True support of ES6 modules
- Optimization is feasible due to explicit `import` definitions
  - Native JS code can also be optimized via DCE
- Scala.js exports are available to JS modules in the bundling process
- No need for JS tools
  - Should have much better performance
- Allows incremental builds
- Backwards compatible (if optional)

**Cons**
- Bundling must be performed to actually use the JS generated by Scala.js
- CommonJS modules must be converted to ES6 before bundling
- Use of GCC is challenging
- Lot of work to implement everything :)

### PR DESCRIPTION
**Pull request by sjrd**
This commit adds a third `ModuleKind` for ES modules, namely `ModuleKind.ESModule`. When emitting an ES module, `@JSImport`s and `@JSExportTopLevel`s straightforwardly map to ES `import` and `export` clauses, respectively.

At the moment, imports are always implemented using a namespace import, then selecting fields inside the namespace. This is suboptimal because it can prevent advanced DCE across ES modules. Improving on this is left for future work.

The Node.js-based environment is adapted to interpret files whose name ends with `.mjs` as ES modules rather than scripts. This aligns with how Node.js itself identifies ES modules as of version 10.x, although it is still experimental, so that could change in the future.

For the 0.6.x branch, the `TestAdapter` assumes that it can force the `JSEnv` to interpret its launcher as an ES module by giving it the `.mjs` extension as well. This is wrong in general, but there does not seem to be a good way to deal with this issue. In 1.x, this will be a non-issue since the `TestAdapter` does not require any launcher.

Although setting `scalaJSLinkerConfig.moduleKind` to `ModuleKind.ESModule` is enough for the Scala.js linker to emit a valid ES module, two additional settings are required to *run* or *test* using Node.js:
```scala
artifactPath in (proj, Compile, fastOptJS) :=
  (crossTarget in (proj, Compile)).value / "somename.mjs"

jsEnv := {
  new org.scalajs.jsenv.NodeJSEnv(
      org.scalajs.jsenv.NODEJSEnv.Config()
        .withArguments(List("--experimental-modules"))
  )
}
```
The first setting is necessary to give the `.mjs` extension to the file produced by Scala.js, which in turn is necessary for Node.js to accept it as an ES module.

The second setting will be necessary until Node.js declares its support for ES module as non-experimental.

ES modules are incompatible with a few features, which are all gone in Scala.js 1.x:

* `EnvironmentInfo.exportsNamespace`: an ES module cannot read its   own exports namespace, short of importing itself (which requires   it to know its file name). The value of `exportsNamespace` is `undefined` in an ES module.
* Generation of a `main` launcher script by the sbt plugin. Attempting to use one will throw an exception in the build.

Moreover, the version of the Closure Compiler that we use does not support ES modules yet, so we deactivate GCC when emitting an ES module.

At this point, the emission of ES modules can be considered stable, but the support in `NodeJSEnv` is experimental (since the support of ES modules in Node.js is itself experimental).

Running the full test suite with ES modules requires Node.js 10.2.0 or later. It has been tested with v10.12.0.

### CODE REVIEW

**Comment by sjrd** [comment]
Looking at the mess I made of `Emitter` and `ClassEmitter`, I'm thinking we should perhaps deprecate `@JSExportTopLevel` with `.`'s inside, i.e., with qualified names. The automatic creation of namespaces is painful, and does not have any direct equivalent in ES. The difference of treatment for `TopLevelFieldExportDef` between unqualified and qualified name is particularly striking me as a bad smell.

**Comment by gzm0** [comment]
Yes, I think deprecating top-level exports with dots is a reasonable idea.

In the worst case, one can always write:

```scala
object Foo {
  @JSExportTopLevel("foo")
  val foo = new js.Object {
    val bar = new js.Object {
      def hello(s: String): String = "Hello " + s
    }
  }
}
```

To prevent too much mess from building up, maybe already disallow this, when exporting to an ESModule?

**Comment by gzm0** [comment]
Is it worth targeting 0.6.x with this? It seems there is a lot of complexity because of old interfaces. (N.B. we do not have `Input` yet).

**Comment by sjrd** [comment]
> To prevent too much mess from building up, maybe already disallow this, when exporting to an ESModule?

Our own test bridge uses a qualified export in 0.6.x :( So I don't think we can disallow it right now. Unless we change the export of the bridge now? Can we do that?

> Is it worth targeting 0.6.x with this? It seems there is a lot of complexity because of old interfaces. (N.B. we do not have Input yet).

I think it is worth it, yes. I was afraid of the complexity initially, but in fact it's not too bad. The main issue is the invocation of the bridge in the test adapter, but it's limited to one method. There is a small price to pay for putting it in 0.6.x, but I believe users will sufficiently benefit from this feature to be worth that price.

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala [hunk 1]
```diff
@@ -134,12 +134,37 @@ abstract class AbstractNodeJSEnv(
...
+        importerFile.content = {
+          s"""
+            |import("${escapeJS(uri)}").catch(e => {
+            |  /* Make sure to fail the process, but give time to Node.js to
+            |   * display a good stack trace before that.
+            |   */
+            |  setTimeout(() => process.exit(1), 100);
+            |  throw e;
+            |});
+          """.stripMargin
```
Comment (gzm0) [inline js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala]: Hum, how is this supposed to work? There is absolutely no guarantee that once this file is required, anything is loaded...

So if there is code like this:
```javascript
require("foo_importer.js");
foo.blah();
```

The second line might randomly fail because `foo` is not loaded yet.

File: test-adapter/src/main/scala/org/scalajs/sbttestadapter/TestAdapter.scala [hunk 2]
```diff
@@ -133,39 +133,62 @@ final class TestAdapter(jsEnv: ComJSEnv, config: TestAdapter.Config) {
...
+            "use strict";
+            $startBridgeFun
+            startBridge(typeof(org) != 'undefined' ? org : {});
+          })();
+        """
+
+      case ModuleKind.ESModule =>
+        val moduleIdent = config.moduleIdentifier.getOrElse {
+          throw new IllegalArgumentException(
+              "The module identifier must be specified for CommonJS modules")
```
Comment (gzm0) [inline test-adapter/src/main/scala/org/scalajs/sbttestadapter/TestAdapter.scala]: ES / EcmaScript?

File: test-suite/js/src/test/require-modules/org/scalajs/testsuite/jsinterop/ModulesTest.scala [hunk 3]
```diff
@@ -26,7 +26,7 @@ class ModulesTest {
 
   @Test def testImportModuleItself(): Unit = {
     val qs = QueryString
-    assertTrue(qs.isInstanceOf[js.Object])
+    assertEquals("object", js.typeOf(qs))
```
Comment (gzm0) [inline test-suite/js/src/test/require-modules/org/scalajs/testsuite/jsinterop/ModulesTest.scala]: Why is this necessary?

File: tools/jvm/src/main/scala/org/scalajs/core/tools/linker/backend/closure/ClosureLinkerBackend.scala [hunk 4]
```diff
@@ -58,8 +58,8 @@ final class ClosureLinkerBackend(
   val symbolRequirements: SymbolRequirement = emitter.symbolRequirements
 
   private val needsIIFEWrapper = moduleKind match {
-    case ModuleKind.NoModule       => true
-    case ModuleKind.CommonJSModule => false
+    case ModuleKind.NoModule                             => true
+    case ModuleKind.ESModule | ModuleKind.CommonJSModule => false
```
Comment (gzm0) [inline tools/jvm/src/main/scala/org/scalajs/core/tools/linker/backend/closure/ClosureLinkerBackend.scala]: I thought closure doesn't support es modules?

File: tools/shared/src/main/scala/org/scalajs/core/tools/javascript/Trees.scala [hunk 5]
```diff
@@ -241,4 +290,53 @@ object Trees {
...
+   *  The `_1` parts of bindings are therefore the identifier names that are
+   *  imported, as specified in `export` clauses of the module. The `_2` parts
+   *  are the names under which they are imported in the current module.
+   *
+   *  Special cases:
+   *  - When `_1 == _2.name`, there is shorter syntax in ES, i.e.,
+   *    `import { binding } from 'from'`.
+   *  - When `_1 == "default"`, it is equivalent to a default import.
+   */
+  case class Import(bindings: List[(String, Ident)], from: StringLiteral)(
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/javascript/Trees.scala]: So `_1` isn't an `Ident` because there are different naming rules I assume? If yes, why don't we need an `Export` (or `ExportIdent`) tree?

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ClassEmitter.scala [hunk 6]
```diff
@@ -1068,31 +1068,38 @@ private[emitter] final class ClassEmitter(jsGen: JSGen) {
...
-    val defProp =
-      genIdentBracketSelect(js.VarRef(js.Ident("Object")), "defineProperty")
-
-    // optional getter definition
-    val getterDef = {
-      js.StringLiteral("get") -> js.Function(Nil, {
-        js.Return(genSelectStatic(cd.encodedName, field))
-      })
-    }
+    if (moduleKind == ModuleKind.ESModule && fullName.indexOf('.') < 0) {
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ClassEmitter.scala]: `!contains('.')`?

File: tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ClassEmitter.scala [hunk 7]
```diff
@@ -1138,9 +1145,16 @@ private[emitter] final class ClassEmitter(jsGen: JSGen) {
    */
   private def genCreateNamespaceInExports(qualName: String)(
       implicit pos: Position): (js.Tree, js.Tree) = {
-    val (createNamespace, namespace, lastPart) =
-      genCreateNamespaceInExportsAndGetNamespace(qualName)
-    (createNamespace, genBracketSelect(namespace, lastPart))
+    if (moduleKind == ModuleKind.ESModule && qualName.indexOf('.') < 0) {
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ClassEmitter.scala]: `!qualName.contains('.')`?

**Comment by gzm0** [comment]
What is our merging strategy for this? It seems it might make sense to mark this `[no-master]` since most of it will not be useful in master (only JS trees from what I see). Especially if we disallow nested top-level exports (which we should IMO, as it seems to be very unidiomatic JS).

**Comment by sjrd** [comment]
> What is our merging strategy for this? It seems it might make sense to mark this [no-master] since most of it will not be useful in master (only JS trees from what I see). Especially if we disallow nested top-level exports (which we should IMO, as it seems to be very unidiomatic JS).

There's a few more things than just the JS trees, e.g., tests, but yeah not much. I've done a test cherry-pick on master, and there are more conflicts than clean merges. It looks like this might be a case where a forward *port* is more appropriate than a forward merge. I'll mark the commit [no-master].

File: js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala (hunk 1 above)
Comment (gzm0) [reply js-envs/src/main/scala/org/scalajs/jsenv/nodejs/AbstractNodeJSEnv.scala]: Hmmm... I see; ok I guess for the 0.6 series with it's unclear semantics of what files passed to a `JSEnv` mean this is OK.

File: tools/shared/src/main/scala/org/scalajs/core/tools/javascript/Trees.scala [hunk 8]
```diff
@@ -21,6 +21,55 @@ import ir.Position.NoPosition
...
+   *  `Other_ID_Continue`. For example,
+   *  `isValidIdentifierName(0x2118.toChar.toString)` will return `false`
+   *  instead of `true`.
+   *
+   *  In theory, it does not really account for code points with the Unicode
+   *  properties `Pattern_Syntax` and `Pattern_White_Space`, which should be
+   *  rejected. However, with the current version of Unicode (9.0.0), there
+   *  seems to be no such character that would be accepted by this method.
+   */
+  final def isValidIdentifierName(name: String): Boolean = {
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/javascript/Trees.scala]: Now this can be private, no?

Comment (sjrd) [reply tools/shared/src/main/scala/org/scalajs/core/tools/javascript/Trees.scala]: It technically can, but I would leave it public. A user of the API would potentially want to call this method to validate inputs, rather than crashing when trying to construct an `ExportName`.

Maybe expose it as `ExportName.isValidExportName` instead?

Comment (gzm0) [reply tools/shared/src/main/scala/org/scalajs/core/tools/javascript/Trees.scala]: That's fair. Yes I like that name better even if it matches less ES terminology. But in terms of our interface it is definitely cleaner.
